lib/widgets.py

```diff
@@ -0,0 +1,102 @@
+      "search", "field" and "values" GET query parameters since those are used
+      in queries.
+    """
+    template_name = 'ajax_search_select.html'
+
+    def __init__(
+            self,
+            ajax: bool = False,
+            display_fields: Optional[List[str]] = None,
+            clipboard_fields: Optional[List[str]] = None,
+            field_sources: Optional[Dict[str, str]] = None,
+            field_labels: Optional[Dict[str, str]] = None,
+            search_api_url: Optional[str] = None,
+            *args: Any,
+            **kwargs: Any,
+            ) -> None:
+        super().__init__(*args, **kwargs)
+        self.ajax = ajax
+        self.display_fields = display_fields or []
+        self.clipboard_fields = clipboard_fields or []
+        self.field_sources = field_sources or {}
+        self.field_labels = field_labels or {}
+        self.search_api_url = search_api_url
+
+    def get_context(self, name: str, value: Any, attrs: Optional[Dict[str, Any]]) -> Dict[str, Any]:
+        """
+        Get data used to render the widget.
+        Overrides `Widget.get_context`.
+        """
+        context = super().get_context(name, value, attrs)
+        context['widget']['attrs'].update({
+            'class': 'search-select-ajax' if self.ajax else 'search-select',
+            'data-display-fields': ','.join(self.display_fields),
+            'data-clipboard-fields': ','.join(self.clipboard_fields),
+            'data-search-api-url': self.search_api_url or '',
+        })
+        context['clipboard_options'] = [
+            {'field': field, 'label': self.field_labels.get(field, field)}
+            for field in self.clipboard_fields
+        ]
+        return context
+
+    def create_option(self, name: str, value: Any, *args: Any, **kwargs: Any) -> Dict[str, Any]:
+        """
+        Get the data used to render a single option in the widget.
+        Overrides `ChoiceWidget.create_option`.
+        """
+        result = super().create_option(name, value, *args, **kwargs)
+
+        # Replace the label with a custom one, if display fields are provided.
+        if self.display_fields:
+            result['label'] = ", ".join((
+                self.get_field_value(value.instance, field)
+                for field in self.display_fields
+            ))
+
+        # Add the data-* attributes that are used for copy/paste functionality.
+        result['attrs'].update({
+            f'data-{field}': self.get_field_value(value.instance, field)
+            for field in self.clipboard_fields
+        })
+
+        return result
+
+    def get_field_value(self, instance: Any, field: str) -> str:
+        """
+        Get the value of a field in an object as a string.
+        """
+        if field in self.field_sources:
+            # Use the same syntax as Django templates for the field source.
+            value = Variable(self.field_sources[field]).resolve(instance)
```

Comment:
As seen here, I've used the `Variable` class, which Django uses to resolve variable references inside templates, such as `{{ profile.user.get_full_name }}`. This allows us to use the same syntax as Django templates when defining data sources for fields (see `UsersSearchSelect` in `lib/fields.py` for an example of this). The other option would have been to pass lambdas in the `field_sources` initialization parameter, or perhaps an instance of a class that has a method which fetches the necessary data.

Reply:
> The other option would have been to pass lambdas in the field_sources initialization parameter, or perhaps an instance of a class that has a method which fetches the necessary data.

Or parse the source string by hand (split at '.' and resolve each part individually) instead of using the Variable class.

Reply:
> I am surprised that this kind of complexity is needed here, but that must be the consequence of a very generalized solution that isn't tied to only Users.

Indeed, it is because I wanted a generalized solution, and UsersSearchSelect uses fields and methods that are not directly available in the UserProfile model (`user.get_full_name`, `user.get_full_name`). So there had to be a way to configure them when instantiating the widget. An alternative option that I had in the code for a while was that the SearchSelect class would have a method (something like `get_data`) that subclasses such as UsersSearchSelect could override, and that method would be used to retrieve the extra fields that aren't directly available in the model.

Reply:
Hmm, I guess there could be yet another option: do not create the labels or populate the data-* attributes in the Python code at all. Instead, let the Javascript code do it by fetching the data from the API (like it already does when searching and selecting users). Then we wouldn't need the `field_sources` at all in the widget initializer.

Reply:
> Hmm, I guess there could be yet another option: do not create the labels or populate the data-* attributes in the Python code at all. Instead, let the Javascript code do it by fetching the data from the API (like it already does when searching and selecting users). Then we wouldn't need the field_sources at all in the widget initializer.

Doesn't seem to be worth the effort. Let's go with the current solution.